### PR TITLE
Worker node taints

### PR DESCRIPTION
**Pull request by danbudris**
*Issue #, if available:*
https://github.com/aws/eks-anywhere/issues/189

*Description of changes:*

This PR will act as a continuation of https://github.com/aws/eks-anywhere/pull/693

dependent on Node Group Configuration [PR](https://github.com/aws/eks-anywhere/pull/968/files):
- [x] index `WorkerNodeGroupConfigurationsSliceEqual` taint comparison using `WorkerNodeGroupConfiguration` `name`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by danbudris** [comment]
https://github.com/mrajashree/cluster-api/pull/5

File: pkg/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -538,6 +538,9 @@ func NeedsNewWorkloadTemplate(oldSpec, newSpec *cluster.Spec, oldVdc, newVdc *v1
 	if oldSpec.Bundles.Spec.Number != newSpec.Bundles.Spec.Number {
 		return true
 	}
+	if !v1alpha1.WorkerNodeGroupConfigurationSliceTaintsEqual(oldSpec.Spec.WorkerNodeGroupConfigurations, newSpec.Spec.WorkerNodeGroupConfigurations) {
```
Comment (g-gaston) [inline pkg/providers/vsphere/vsphere.go:541]: Do we need this for docker or other providers too?

File: pkg/validations/cluster.go [hunk 2]
```diff
@@ -9,9 +9,30 @@ import (
 
 func ValidateTaintsSupport(clusterSpec *cluster.Spec) error {
 	if !features.IsActive(features.TaintsSupport()) {
-		if len(clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Taints) > 0 {
+		wngcTaintsPresent := false
```
Comment (g-gaston) [inline pkg/validations/cluster.go:12]: - [ ] Ticket to come back and cleanup these `wngc` names

File: pkg/validations/cluster.go [hunk 3]
```diff
@@ -9,9 +9,30 @@ import (
...
+				wngcTaintsPresent = true
+				break
+			}
+		}
+
+		if len(clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Taints) > 0 ||
+			wngcTaintsPresent {
 			return fmt.Errorf("Taints feature is not enabled. Please set the env variable TAINTS_SUPPORT.")
 		}
+	} else if len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].Taints) > 0 {
```
Comment (g-gaston) [inline pkg/validations/cluster.go:24]: I think we need to change/remove this. It doesn't really work with multiple worker node groups

Also, this should probably have unit tests, but that could be in a separate ticket

File: pkg/providers/vsphere/vsphere.go (hunk 1 above)
Comment (danbudris) [reply pkg/providers/vsphere/vsphere.go:541]: yeah, it'll be added to other providers; I actually just pushed that a few minutes ago after testing it out on vsphere and confirming the behavior (e.g. it causes an upgrade to happen and the nodes to roll, which it does)

File: pkg/validations/cluster.go (hunk 2 above)
Comment (danbudris) [reply pkg/validations/cluster.go:12]: - [ ] test

**Comment by danbudris** [comment]
https://github.com/aws/eks-anywhere/issues/1125

File: pkg/providers/docker/docker.go [hunk 4]
```diff
@@ -163,6 +163,7 @@ func (d *DockerTemplateBuilder) GenerateCAPISpecWorkers(clusterSpec *cluster.Spe
 		} else {
 			values["workloadTemplateName"] = d.WorkerMachineTemplateName(clusterSpec.Name, workerNodeGroupConfiguration.Name)
 		}
+		values["workerNodeGroupTaints"] = workerNodeGroupConfiguration.Taints
```
[the remark this answers is not in the document]
Comment (danbudris) [reply pkg/providers/docker/docker.go:166]: addressed

File: pkg/providers/vsphere/vsphere.go (hunk 1 above)
Comment (danbudris) [reply pkg/providers/vsphere/vsphere.go:541]: done for docker

File: pkg/validations/cluster.go (hunk 3 above)
Comment (danbudris) [reply pkg/validations/cluster.go:24]: https://github.com/aws/eks-anywhere/issues/1126

**Review by g-gaston — APPROVED** [review]
/lgtm

**Pull request merged**